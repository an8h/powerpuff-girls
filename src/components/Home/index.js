// @flow

/* 
Home Component: 
It is the first shown page when the user runs the app. 
Rendered UI:
movie      : Shows the Powerpuff movie details
episodes   : Shows a list of episodes seperated by season
*/

import React from "react";
import { connect } from "react-redux";
import { mapDispachToProps } from "../../redux/Store";
import ItemToShow from "../ItemToShow/index";
import Episodes from "../Episodes/index";
import { MOVIE_ID } from "../../config";
import groupEpisodesBySeasons from "../../utils";

type Props = {
  getMovie: Function,
  getEpisodes: Function,
  movie: Object,
  movieIsLoading: Boolean,
  episodes: Object,
  episodesIsLoading: Boolean,
};

class Home extends React.Component<Props> {
  componentDidMount() {
    const { getMovie, getEpisodes } = this.props;

    // Make the API call and get the needed data
    getMovie(MOVIE_ID);
    getEpisodes(MOVIE_ID);
  }

  render() {
    const { movie, movieIsLoading, episodes, episodesIsLoading } = this.props;
    const episodesBySeasons = groupEpisodesBySeasons(episodes);

    return (
      <div>
        <ItemToShow item={movie} />
        {Object.keys(episodesBySeasons).map((key) => {
          return (
            <div key={key}>
              <h1>Season {key}</h1>
              <Episodes key={key} episodes={episodesBySeasons[key]} />
            </div>
          );
        })}
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    ...ownProps,
    movie: state.getMovie.data,
    movieIsLoading: state.getMovie.isLoading,
    episodes: state.getEpisodes.data,
    episodesIsLoading: state.getEpisodes.isLoading,
  };
};

export default connect(mapStateToProps, mapDispachToProps)(Home);
