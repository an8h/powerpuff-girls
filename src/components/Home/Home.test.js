import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import Home from "./index";

const mockStore = configureStore([]);

describe("Home Component", () => {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore({
      movie: {
        data: [],
        isLoading: false,
        errors: [],
      },
      episodes: {
        data: [],
        isLoading: false,
        errors: [],
      },
    });

    store.dispatch = jest.fn();

    component = renderer.create(
      <Provider store={store}>
        <Home />
      </Provider>
    );
  });

  it("should render with given state from Redux store", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});
