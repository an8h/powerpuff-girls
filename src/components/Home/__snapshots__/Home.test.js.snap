// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component should render with given state from Redux store 1`] = `
<div>
  <div />
</div>
`;
